rust/src/id.rs

```diff
@@ -0,0 +1,128 @@
+use std::str::FromStr;
+
+use percent_encoding::{percent_decode, percent_encode, NON_ALPHANUMERIC};
+
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub struct FullId {
+    pub scheme: String,
+    pub id: String,
+    pub path: String,
+}
+
+impl FullId {
+    pub fn new(scheme: String, id: String, path: String) -> Self {
+        FullId { scheme, id, path }
+    }
+
+    pub fn split(self) -> (BackendId, String) {
+        return (BackendId::new(self.scheme, self.id), self.path);
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub struct BackendId {
+    pub scheme: String,
+    pub id: String,
+}
+
+impl BackendId {
+    pub fn new(scheme: String, id: String) -> Self {
+        BackendId { scheme, id }
+    }
+
+    pub fn relative(&self, path: String) -> FullId {
+        FullId::new(self.scheme.clone(), self.id.clone(), path)
+    }
+
+    pub fn user_to_relative(&self, user: crate::proto::User) -> crate::proto::User {
+        crate::proto::User {
+            id: self.relative(user.id).to_string(),
+            display_name: user.display_name,
+        }
+    }
+
+    pub fn source_to_relative(
+        &self,
+        source: crate::proto::ChannelSource,
+    ) -> crate::proto::ChannelSource {
+        crate::proto::ChannelSource {
+            channel_id: self.relative(source.channel_id).to_string(),
+            user: source.user.map(|user| self.user_to_relative(user)),
+        }
+    }
+}
```

Comment:
This feels like it's defined the wrong way to me. Instead, can you define a trait and then add it onto `ChannelSource`?

```rust
trait ChannelSourceToRelative {
    fn to_relative(&self, backend_id: &BackendId) -> crate::proto::ChannelSource;
}

impl ChannelSourceToRelative for crate::proto::ChannelSource {
    fn to_relative(&self, backend_id: &BackendId) {
        crate::proto::ChannelSource {
            channel_id: backend_id.relative(self.channel_id).to_string(),
            user: self.user.map(|user| user.to_relative(backend_id)),
        }
    }
}
```

Not sure how you feel about this. This would require doing the same thing for `user_to_relative`. To me, both of these operate on the proto objects and not on the ID, and instead just use the ID for information.

Reply:
I agree with you in theory, but as far as I know we can't attach additional methods to the generated protobuf objects.

Reply:
If you define the trait then you can impl the trait on external structs

Reply:
Turns out you can actually declare methods directly on proto items by doing it where you call include_proto. So, I did that... there's now an into_relative method on both user and ChannelSource.